lib/fs.js

```diff
@@ -964,83 +1020,101 @@ fs.statSync = function(path) {
+                               ['string', 'Buffer', 'URL']);
+  }
   return binding.readlink(pathModule._makeLong(path), options.encoding);
 };
 
-function preprocessSymlinkDestination(path, type, linkPath) {
-  if (!isWindows) {
-    // No preprocessing is needed on Unix.
-    return path;
-  } else if (type === 'junction') {
-    // Junctions paths need to be absolute and \\?\-prefixed.
-    // A relative target is relative to the link's parent directory.
-    path = pathModule.resolve(linkPath, '..', path);
-    return pathModule._makeLong(path);
-  } else {
-    // Windows symlinks don't tolerate forward slashes.
-    return ('' + path).replace(/\//g, '\\');
+fs.symlink = function(target, path, type, callback) {
+  if (typeof type === 'function') {
+    callback = type;
+    type = 'file';
   }
-}
+  const _type = stringToSymlinkType(type);
 
-fs.symlink = function(target, path, type_, callback_) {
-  var type = (typeof type_ === 'string' ? type_ : null);
-  var callback = makeCallback(arguments[arguments.length - 1]);
+  handleError(target = getPathFromURL(target));
+  handleError(path = getPathFromURL(path));
 
-  if (handleError((target = getPathFromURL(target)), callback))
-    return;
+  nullCheck(target);
+  nullCheck(path);
 
-  if (handleError((path = getPathFromURL(path)), callback))
-    return;
+  if (typeof target !== 'string' && !Buffer.isBuffer(target)) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'target',
+                               ['string', 'Buffer', 'URL']);
+  }
 
-  if (!nullCheck(target, callback)) return;
-  if (!nullCheck(path, callback)) return;
+  if (typeof path !== 'string' && !Buffer.isBuffer(path)) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'path',
+                               ['string', 'Buffer', 'URL']);
+  }
 
-  var req = new FSReqWrap();
-  req.oncomplete = callback;
+  const req = new FSReqWrap();
+  req.oncomplete = makeCallback(callback);
 
   binding.symlink(preprocessSymlinkDestination(target, type, path),
                   pathModule._makeLong(path),
-                  type,
+                  _type,
                   req);
 };
 
-fs.symlinkSync = function(target, path, type) {
-  type = (typeof type === 'string' ? type : null);
-  handleError((target = getPathFromURL(target)));
-  handleError((path = getPathFromURL(path)));
+fs.symlinkSync = function(target, path, type = 'file') {
+  const _type = stringToSymlinkType(type);
+  handleError(target = getPathFromURL(target));
+  handleError(path = getPathFromURL(path));
   nullCheck(target);
   nullCheck(path);
 
+  if (typeof target !== 'string' && !Buffer.isBuffer(target)) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'target',
+                               ['string', 'Buffer', 'URL']);
+  }
+
+  if (typeof path !== 'string' && !Buffer.isBuffer(path)) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'path',
+                               ['string', 'Buffer', 'URL']);
```

Comment:
I think it would be great if all the errors could also contain the actual value.